torchci/lib/bot/greenBot.ts

```diff
@@ -0,0 +1,231 @@
+      // If all jobs are successful or neutral, then try merge, else do nothing
+      if (successfulJobs.length === jobs.length) {
+        await tryLandPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "All checks passed. Attempting to merge.",
+        });
+      }
+      // If any failed, alert the author
+      else if (failedJobs.length > 0) {
+        await failPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "Failed to land PR due to red signal.",
+        });
+      }
+    }
+  });
+
+  app.on("workflow_job.completed", async (ctx) => {
+    const rocksetClient = getRocksetClient();
+    // Get all PRs that have this as the head SHA and the land pending label
+    const landPendingPrQuery =
+      await rocksetClient.queryLambdas.executeQueryLambda(
+        "commons",
+        "prs_with_label",
+        rocksetVersions.commons.prs_with_label,
+        {
+          parameters: [
+            {
+              name: "label",
+              type: "string",
+              value: LAND_PENDING,
+            },
+            {
+              name: "sha",
+              type: "string",
+              value: ctx.payload.workflow_job.head_sha,
+            },
+          ],
+        }
+      );
+
+    const prs = landPendingPrQuery?.results;
+    if (prs == null || prs?.length === 0) {
+      return;
+    }
+
+    if (prs.length > 1) {
+      // If there's more than 1 PR with the same sha and the land-pending label,
+      // Then we don't want to potentially race.
+      return;
+    }
+
+    const landPendingPr = prs[0];
+
+    const jobs = await getJobsBySha(landPendingPr.sha);
+    const [successfulChecks, failedChecks] = categorizeJobs(jobs);
+
+    const conclusion = ctx.payload.workflow_job.conclusion;
+
+    const prData = {
+      octokit: ctx.octokit,
+      ...getPayloadData(ctx),
+      issue_number: landPendingPr.number,
+    };
+    if (conclusion == "failure") {
+      // Only alert if this is the first failure
+      if (failedChecks.length < 1) {
+        await failPR({
+          ...prData,
+          comment: "Failed to land PR due to failing signal",
+        });
+      }
+    } else if (successfulChecks.length === jobs.length) {
+      await tryLandPR({
+        ...prData,
+        comment: "All checks passed. Attempting to merge.",
+      });
+    }
+  });
+}
+
+async function getJobsBySha(sha: string) {
+  const rocksetClient = getRocksetClient();
+
+  const jobsQuery = await rocksetClient.queryLambdas.executeQueryLambda(
+    "commons",
+    "workflow_jobs_for_sha",
+    rocksetVersions.commons.workflow_jobs_for_sha,
+    {
+      parameters: [
+        {
+          name: "sha",
+          type: "string",
+          value: sha,
+        },
+      ],
+    }
+  );
+  return jobsQuery.results ?? [];
+}
+
+function categorizeJobs(jobs: any[]) {
```

Comment:
In the future, this should only check against required status checks

Reply:
I think Michael preferred to do jobs here. Also when I queried the checks using the GH API, it always returned an empty array or was stuck on pending.